src/main/java/org/opensearch/security/configuration/SecurityConfigVersionHandler.java

```diff
@@ -0,0 +1,330 @@
+
+            log.info("Successfully saved version {} to {}", version.getVersion_id(), SecurityConfigVersionsIndex);
+
+            // Async retention task
+            threadPool.generic().submit(() -> {
+                try {
+                    applySecurityConfigVersionIndexRetentionPolicy();
+                } catch (Exception e) {
+                    log.warn("Retention policy async failed", e);
+                }
+            });
+
+        } catch (org.opensearch.index.engine.VersionConflictEngineException conflictEx) {
+            log.warn("Concurrent update detected on {}", SecurityConfigVersionsIndex);
+        } catch (Exception e) {
+            log.error("Failed to save version to {}", SecurityConfigVersionsIndex, e);
+            throw ExceptionsHelper.convertToOpenSearchException(e);
+        }
+    }
+
+    private boolean shouldSkipVersionUpdate(SecurityConfigVersionDocument document, SecurityConfigVersionDocument.Version<?> newVersion) {
+        SecurityConfigVersionsLoader.sortVersionsById(document.getVersions());
+
+        if (!document.getVersions().isEmpty()) {
+            SecurityConfigVersionDocument.Version<?> latestVersion = document.getVersions().get(document.getVersions().size() - 1);
+            Map<String, SecurityConfig<?>> latestConfigMap = latestVersion.getSecurity_configs();
+            Map<String, SecurityConfig<?>> newConfigMap = newVersion.getSecurity_configs();
+
+            if (!SecurityConfigDiffCalculator.hasSecurityConfigChanged(latestConfigMap, newConfigMap)) {
+                log.info("No changes detected in security configuration. Skipping version update.");
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private void writeSecurityConfigVersion(SecurityConfigVersionDocument document, long currentSeqNo, long currentPrimaryTerm)
+        throws IOException {
+        Map<String, Object> updatedDocMap = document.toMap();
+        String json = DefaultObjectMapper.objectMapper.writeValueAsString(updatedDocMap);
+
+        var indexRequest = new org.opensearch.action.index.IndexRequest(SecurityConfigVersionsIndex).id(
+            "opendistro_security_config_versions"
+        ).source(json, XContentType.JSON).setRefreshPolicy(RefreshPolicy.IMMEDIATE);
+
+        if (currentSeqNo >= 0 && currentPrimaryTerm > 0) {
+            indexRequest.setIfSeqNo(currentSeqNo);
+            indexRequest.setIfPrimaryTerm(currentPrimaryTerm);
+        }
+
+        client.index(indexRequest).actionGet();
+    }
+
+    public void applySecurityConfigVersionIndexRetentionPolicy() {
+        SecurityConfigVersionDocument document = configVersionsLoader.loadFullDocument();
+        List<SecurityConfigVersionDocument.Version<?>> versions = document.getVersions();
+
+        SecurityConfigVersionsLoader.sortVersionsById(versions);
+
+        if (versions.size() > maxVersionsToKeep) {
```

Comment:
Can we do this operation at the same time a new config is added to ensure the list does not exceed the maxVersionsToKeep.

If the list is always sorted we can append on the end and remove from the front simultaneously.